### PR TITLE
remove non-consensus data when inserting new consensus blocks

### PR DESCRIPTION
**Pull request by ayrat555**
In this PR (https://github.com/poanetwork/blockscout/pull/1657) we started invalidating a large number of blocks.

Imagine a situation when we just imported a non-consensus block. Logic in this PR sets `consensus` to false for this block in the DB to make `Catchup.Fetcher` re-fetch this block. But we already imported logs and token transfers for this non-consensus block. the second time we will import correct logs and transfers but invalid data is not deleted.

fixes https://github.com/poanetwork/blockscout/issues/2645
fixes https://github.com/poanetwork/blockscout/issues/2627

## Changelog
- remove non-consensus data when inserting new consensus blocks

### CODE REVIEW

File: apps/explorer/lib/explorer/chain/import/runner/blocks.ex [hunk 1]
```diff
@@ -58,6 +58,31 @@ defmodule Explorer.Chain.Import.Runner.Blocks do
...
+      lose_invalid_neighbour_consensus(repo, where_invalid_neighbour, insert_options)
+    end)
+    |> Multi.run(:remove_nonconsensus_data, fn repo,
+                                               %{
+                                                 lose_consensus: lost_consensus_blocks,
+                                                 lose_invalid_neighbour_consensus: lost_consensus_neighbours
+                                               } ->
+      nonconsensus_block_numbers =
+        lost_consensus_blocks
+        |> Kernel.++(lost_consensus_neighbours)
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/chain/import/runner/blocks.ex]: Just a nitpick, but I don't think `credo` will complain if you start the pipe with `lost_consensus_blocks ++ lost_consensus_neighbours`

File: apps/explorer/lib/explorer/chain/import/runner/blocks.ex [hunk 2]
```diff
@@ -342,6 +361,87 @@ defmodule Explorer.Chain.Import.Runner.Blocks do
...
+      {:ok, %{token_transfers: deleted_token_transfers, logs: deleted_logs}}
+    end
+  end
+
+  defp remove_nonconsensus_token_transfers(repo, nonconsensus_block_numbers, %{timeout: timeout}) do
+    ordered_token_transfers =
+      from(token_transfer in TokenTransfer,
+        where: token_transfer.block_number in ^nonconsensus_block_numbers,
+        select: map(token_transfer, [:transaction_hash, :log_index]),
+        order_by: [
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/chain/import/runner/blocks.ex]: Can you please add the line `# Enforce TokenTransfer ShareLocks order (see docs: sharelocks.md)` before this one?

File: apps/explorer/lib/explorer/chain/import/runner/blocks.ex [hunk 3]
```diff
@@ -342,6 +361,87 @@ defmodule Explorer.Chain.Import.Runner.Blocks do
...
+
+  defp remove_nonconsensus_logs(repo, nonconsensus_block_numbers, %{timeout: timeout}) do
+    transaction_query =
+      from(transaction in Transaction,
+        where: transaction.block_number in ^nonconsensus_block_numbers,
+        select: map(transaction, [:hash]),
+        order_by: transaction.hash
+      )
+
+    ordered_logs =
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/chain/import/runner/blocks.ex]: This query needs to be modified. By making a `join` with the `transaction_query` your `lock: "FOR UPDATE"` will also lock the rows in the `transactions` table, but you don't need to modify them.

You can solve this by using `lock: "FOR UPDATE OF l0"` instead, see [this example](https://github.com/poanetwork/blockscout/blob/pp-enforce-sharelock-order/apps/explorer/lib/explorer/chain/import/runner/blocks.ex#L598).
You can also avoid the `transaction_query` and move the condition to after the `join` in this one.

File: apps/explorer/lib/explorer/chain/import/runner/blocks.ex (hunk 2 above)
Comment (ayrat555) [reply apps/explorer/lib/explorer/chain/import/runner/blocks.ex]: done. Also, I added the same comment for logs removal query

File: apps/explorer/lib/explorer/chain/import/runner/blocks.ex (hunk 3 above)
Comment (ayrat555) [reply apps/explorer/lib/explorer/chain/import/runner/blocks.ex]: I added `"FOR UPDATE OF l0"`

**Pull request merged**